packages/nimble-components/src/table/specs/action-menu-hld.md

```diff
@@ -0,0 +1,113 @@
+
+## Implementation / Design
+
+### Client-facing API
+
+There are two pieces of configuration that must be provided by a client in order to use the action menu:
+
+1. Specifying which column, or columns, the action menu will be visible within
+2. Specifying the menu and its items to show when the menu button is open
+
+The nimble design system will be opinionated about many details of the menu button within a table. Therefore, a client will not be able to configure the exact details of the menu button itself, such as:
+
+-   The appearance mode of the button
+-   The icon shown within the button
+-   When the button becomes visible (e.g. always visible, visible on hover/selection only, etc.)
+
+If the need arises for a client to have more control of the menu button's configuration, the API can be extended at a later time.
+
+#### Specifying a column to include a menu button
+
+A client can specify that a column should have an action menu within it by adding the `show-menu` attribute on the column definition. In the example below, the menu will be added to the _First name_ column only. The `show-menu` attribute can be added to multiple columns if multiple action menus are desired within a row.
+
+```HTML
+<nimble-table>
+    <nimble-table-column-text field-name="firstName" show-menu>First name</nimble-table-column-text>
+    <nimble-table-column-text field-name="lastName">Last name</nimble-table-column-text>
+</nimble-table>
+```
+
+### Providing a menu
+
+Because only one menu can be open on the page at a time, the client can provide a single menu that will be associated with the correct menu button when it is open. The menu is provided by the client by slotting a menu in the `action-menu` slot as shown below:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-text field-name="firstName" show-menu>First name</nimble-table-column-text>
+    <nimble-table-column-text field-name="lastName">Last name</nimble-table-column-text>
+
+    <nimble-menu slot="action-menu">
+        <nimble-menu-item>My first action</nimble-menu-item>
+        <nimble-menu-item>My second action</nimble-menu-item>
+        <nimble-menu-item>My last action</nimble-menu-item>
+    </nimble-menu>
+</nimble-table>
+```
+
+If an application requires different menu items for different rows, the client is responsible for ensuring that the items in the menu are correct for the row(s) that the menu is associated. This can be done by handling the `action-menu-opening` event on the table and updating the menu items as appropriate. The `action-menu-opening` event will include the following in its details:
+
+-   `rowIds` - string array - The IDs of the rows that the menu is associated with.
+-   `columnId` - string, possibly undefined - The ID of the column that the menu is associated with. A column ID is optional on a column definition. If the menu is associated with a column without an ID, `columnId` will be `undefined` in the event details.
+
```

Comment:
Could you provide a short section that describes how the client would respond to a menu item being selected? How would they know which table row/column it was associated with? Would they have to store that state themselves in the handler for `action-menu-opening` or is it retrievable from the menu's `change` event data somehow?

Reply:
I added a section saying that the data from the `action-menu-opening` event should be cached to have the appropriate context for handling events coming from the menu.